src/algebra/quandle.lean

```diff
@@ -0,0 +1,586 @@
+mathematicians.  David Joyce introduced quandles in his thesis
+[Joyce1982] to define an algebraic invariant of knot and link
+complements that is analogous to the fundamental group of the
+exterior, and he showed that the quandle associated to an oriented
+knot is invariant up to orientation-reversed mirror image.  Racks were
+used by Fenn and Rourke for framed codimension-2 knots and
+links.[FennRourke1992]
+
+The name "rack" came from wordplay by Conway and Wraith for the "wrack
+and ruin" of forgetting everything but the conjugation operation for a
+group.
+
+## Main definitions
+
+* `shelf` is a type with a self-distributive action
+* `rack` is a shelf whose action for each element is invertible
+* `quandle` is a rack whose action for an element fixes that element
+* `quandle.conj` defines a quandle of a group acting on itself by conjugation.
+* `shelf_hom` is homomorphisms of shelves, racks, and quandles.
+* `rack.envel_group` gives the universal group the rack maps to as a conjugation quandle.
+* `rack.opp` gives the rack with the action replaced by its inverse.
+
+## Main statements
+* `rack.envel_group` is universal (`to_envel_group.univ` and `to_envel_group.univ_uniq`)
+
+## Notation
+* `x ◃ y` is local notation for `shelf.act x y`
+* `x ◃⁻¹ y` is local notation for `rack.inv_act x y`
+* `S →◃ S'` is local notation for `shelf_hom S S'`
+
+## Todo
+
+* If g is the Lie algebra of a Lie group G, then (x ◃ y) = Ad (exp x) x forms a quandle
+* If X is a symmetric space, then each point has a corresponding involution that acts on X, forming a quandle.
+* Alexander quandle with `a ◃ b = t * b + (1 - t) * b`, with `a` and `b` elements of a module over Z[t,t⁻¹].
+* If G is a group, H a subgroup, and z in H, then there is a quandle `(G/H;z)` defined by
+  `yH ◃ xH = yzy⁻¹xH`.  Every homogeneous quandle (i.e., a quandle Q whose automorphism group acts
+  transitively on Q as a set) is isomorphic to such a quandle.
+  There is a generalization to this arbitrary quandles in [Joyce's paper (Theorem 7.2)][Joyce1982].
+
+## Tags
+
+rack, quandle
+-/
+open opposite
+
+universes u v
+
+/--
+A *shelf* is a structure with a self-distributive binary operation.
+The binary operation is regarded as a left action of the type on itself.
+-/
+class shelf (α : Type u) :=
+(act : α → α → α)
+(self_distrib : ∀ {x y z : α}, act x (act y z) = act (act x y) (act x z))
+
+/--
+A *rack* is an automorphic set (a set with an action on itself by
+bijections) that is self-distributive.  It is a shelf such that each
+element's action is invertible.
+
+The local notations `x ◃ y` and `x ◃⁻¹ y` denote the action and the
+inverse action, respectively, and they are right associative.
+-/
+class rack (α : Type u) extends shelf α :=
+(inv_act : α → α → α)
+(left_inv : ∀ x, function.left_inverse (inv_act x) (act x))
+(right_inv : ∀ x, function.right_inverse (inv_act x) (act x))
+
+local infixr ` ◃ ` : 65 := shelf.act
+local infixr ` ◃⁻¹ ` : 65 := rack.inv_act
```

Comment:
I guess one is meant to create a `locale` so other people can quickly reintroduce the same notions. I don't remember off-hand how this works, but there must be examples! Could you try this?

Reply:
Done!